### PR TITLE
Add ARCHITECTURE.md

### PR DESCRIPTION
**Pull request by thomaseizinger**

### CODE REVIEW

**Review by da-kami — APPROVED** [review]

**Comment by da-kami** [comment]
The vision on the library section being a set of pure functions but not a state machine is interesting, we will keep that in mind in our further design for the MVP.

File: docs/ARCHITECTURE.md [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+To keep our UI stateless and responsive, we always update the local state of our daemon first (and emit an event for it).
+Only once the local state is updated, we engage with other systems like the maker/taker daemon.
+
+Applying state changes locally first allows us to record the user's intention and provide instant (UI) feedback that we are working on making it happen.
+Even if the user restarts the entire application, we can pick up where we left of and finish what we were meant to be doing.
+
+### Append only database
+
+The database structure is append only.
+We never update / overwrite existing state to make sure we don't ever loose data due to bugs in state transitions.
```
Comment (luckysori) [inline docs/ARCHITECTURE.md]: Sorry, I can't help myself :see_no_evil: 

```suggestion
We never update / overwrite existing state to make sure we don't ever lose data due to bugs in state transitions.
```

File: docs/ARCHITECTURE.md [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+
+### Library only exposes pure transition functions rather than a state machine
+
+The protocol implemented in the library can be thought of as a state machine that is pushed forward by each party.
+To remain flexible in how the protocol is used, the library MUST only expose pure functions to go from one state to the other rather than representing the actual states itself.
+This allows applications on top of shape their states and messages as they wish, only reaching into the library for doing the heavy lifting of cryptography and other protocol-specific functionality.
```
Comment (luckysori) [inline docs/ARCHITECTURE.md]: :wrench: 

```suggestion
This allows applications on top to shape their states and messages as they wish, only reaching into the library for doing the heavy lifting of cryptography and other protocol-specific functionality.
```

**Review by luckysori — APPROVED** [review]
Thanks for writing this! 

Just a couple of minor fixes.

**Pull request merged**